tests/diagnostics/non-void-return-constructor.slang

```diff
@@ -0,0 +1,47 @@
+//TEST:SIMPLE(filecheck=CHECK):
+
+// Test to catch compilation time errors with return values in a constructor / member functions within a class / struct.
+struct MyType
+{
+    float v;
+
+    // A constructor from the callsite's point of view is a function that returns MyType.
+    // A constructor from the inside the body is treated as a function that modifies `this`.
+    // because we are expected to write:
+    // this.v = 5;
+
+    // A test case where the constructor returns a value.
+    // This automatically gets resolved as this = t; return this;
+    __init()
+    {
+        MyType t;
+        t.v = 5;
+        return t;
+    }
+
+    // A valid use-case when the constructor returns a void
+    __init(int v)
+    {
+        if (v == 0)
+        {
+            this.v = 5;
+            return this;
+        }
+        this.v = v;
+    }
```

Comment:
This code looks little confusing.
Should we allow the function to return a value at the same time it doesn't need to have a return statement at the end?

Reply:
this is a positive test that confirms that it is ok to have a return from the constructor. i will change `return this;` to `return;` and possibly add another `return;` at the end if it helps. or i can just leave `return this;` as it is, and only add another `return;` at the end.